internal/vsock/client.go

```diff
@@ -0,0 +1,230 @@
+// Copyright (c) Ultraviolet
+// SPDX-License-Identifier: Apache-2.0
+
+package vsock
+
+import (
+	"encoding/binary"
+	"fmt"
+	"io"
+	"log"
+	"net"
+	"sync"
+	"time"
+
+	"google.golang.org/protobuf/proto"
+)
+
+const (
+	maxRetries     = 3
+	retryDelay     = time.Second
+	maxMessageSize = 1 << 20 // 1 MB
+	ackTimeout     = 5 * time.Second
+	maxConcurrent  = 100 // Maximum number of concurrent messages
+)
+
+type Message struct {
+	ID      uint32
+	Content []byte
+}
+
+type AckWriter struct {
+	conn            net.Conn
+	pendingMessages chan *Message
+	ackChannels     map[uint32]chan bool
+	ackMu           sync.RWMutex
+	nextID          uint32
+	done            chan struct{}
+	wg              sync.WaitGroup
+}
+
+func NewAckWriter(conn net.Conn) *AckWriter {
+	aw := &AckWriter{
+		conn:            conn,
+		pendingMessages: make(chan *Message, maxConcurrent),
+		ackChannels:     make(map[uint32]chan bool),
+		nextID:          1,
+		done:            make(chan struct{}),
+	}
+	aw.wg.Add(2)
+	go aw.sendMessages()
+	go aw.handleAcknowledgments()
+	return aw
+}
+
+func (aw *AckWriter) WriteProto(msg proto.Message) (int, error) {
+	data, err := proto.Marshal(msg)
+	if err != nil {
+		return 0, fmt.Errorf("error marshaling protobuf message: %v", err)
+	}
+	return aw.Write(data)
+}
+
+func (aw *AckWriter) Write(p []byte) (int, error) {
+	if len(p) > maxMessageSize {
+		return 0, fmt.Errorf("message size exceeds maximum allowed size of %d bytes", maxMessageSize)
+	}
+
+	aw.ackMu.Lock()
+	messageID := aw.nextID
+	aw.nextID++
+	aw.ackMu.Unlock()
+
+	ackCh := make(chan bool, 1)
+	aw.ackMu.Lock()
```

Comment:
Don't unlock. It's more expensive than the allocation for which you are releasing it, it's not worth it.